### ISSUE TITLE
Create sourceItem property

### ISSUE THREAD

**Issue by philbarker** [CredentialEngine/Schema-Development#769]
When pointing to some thing we need a property that has the thing in its range:

Add property: 
name: sourceItem
definition: A resource that is referenced by a pathway component
domain:
range:

**Comment by jeannekitchens** [comment]
@stuartasutton @siuc-nate please review this proposal.  Let's get a final proposal positioned to be implemented.

**Comment by stuartasutton** [comment]
Are we adverse to using `ceterms:source` defined as "[a]uthoritative source of an entity's information."

**Comment by philbarker** [comment]
`ceterms:source` seems different to me, it seems to meet the use case of "here's link where I got the data about this entity", but I think that what is needed here is "here's the entity for which this resource is a proxy that has a certain position in a pathway". It's a peculiar relationship.

**Comment by stuartasutton** [comment]
I.e., `proxyFor`

**Comment by philbarker** [comment]
More specifically: [`ore:proxyFor`](https://www.openarchives.org/ore/1.0/vocabulary#ore-proxyFor)

**Comment by stuartasutton** [comment]
Definition of `ore:proxifyFor` is a bit off: "Proxy objects are used to represent a Resource as it is aggregated in a particular Aggregation."

**Comment by philbarker** [comment]
@stuartasutton I think the second sentence is the one with meaning that we want, "The ore:proxyFor relationship is used to link the proxy to the Aggregated Resource it is a proxy for", though we would want to remove the ORE-specific terminology of Proxy Object, Aggregated Resource and subsititute our own PathwayComponent. And fix the varying number for the nouns. Something like:

Definition: link to the resource that is referenced in a Pathway.
Usage note: PathwayComponents are used to represent resources as they are aggregated in a particular pathway, their relationship to other components of the pathway are not universally true of the resource the represent. The object of the proxyFor property is the resource that is being represented.

(Yes, that's a very long usage note, it's halfway to being a section in the handbook and half of it should also be a usage note for PathwayComponents, but I can't think of anything shorter that doesn't raise more questions than it answers.)

**Comment by siuc-nate** [comment]
Regardless of its definition, our original intent in creating ceterms:sourceData was to reference the **resource**. The only reason it mentions "structured data" is because:
1. We needed to differentiate it from **ceasn**:source, which could be anything (webpage, PDF, etc) and was specific to competency frameworks.  
2. We already had **ceterms**:source, which (as Stuart points out), has a different function based on its definition.

The fact that this _can be read as_ referencing a JSON document representing a resource **instead of the resource itself** is unintentional. I wanted to clarify this background given the misunderstanding about it that we found ourselves working through a few weeks ago.

That said, if we still need to add a new property that does the same thing we created ceterms:sourceData to do, I do prefer "proxy for" over "source item".  We may be able to find some other purpose for sourceData given its (unfortunate) definition.

**Comment by siuc-nate** [comment]
I would propose a definition/usage note that isn't explicitly tied to pathways, given our past experience with needing to come back and loosen up definitions later.  Something like:

**URI:** ceterms:proxyFor
**Label:** Proxy For
**Definition:** Resource that this resource is a reference to or a stand-in for.
**Usage Note:** Assertions made about this proxy resource should not be assumed to necessarily be true of the resource being proxied.

**Comment by stuartasutton** [comment]
Please, do not use `sameAs`.

**Comment by philbarker** [comment]
Thanks @siuc-nate I hink that sourceData should stand as defined, it may be useful in the future.

I think the intent of your definition is fine, but the defintion is a bit tortuous (but then the whole business of proxies is a bit tortuous). Also, I'm increasingly thinking that we should define the property, not the object the property points to (yes I know that you, I and everyone else have been doing that forever). The comment talks about "this proxy resource" which introduces the idea of proxy resources without explaining what it has to do with pathway components, and is about the subject not the use of the property.

How about:

> **URI:** ceterms:proxyFor
> **Label:** Proxy For
> **Definition:** Indicates the resource for which a pathway component or similar proxy resource is a stand-in.
> **Comment:** The object for this property will be acting as proxy for the resource that is its subject in a context such as a pathway so that assertions can be made about relationships between things in that context that are not universially true of the resources for which they are proxies.

We could also add comments to the PathwayComponent classes

**Remove**

> **Subject:** ceterms:PathwayComponent
> **Predicate:** rdfs:comment 
> **Object:** An instance of PathwayComponent may serve as the goal or destination node of a Pathway.

**Add**

> **Subject:** ceterms:PathwayComponent
> **Predicate:** rdfs:comment 
> **Object:** An instance of PathwayComponent may serve as the goal or destination node of a Pathway. A Pathway Component acts as a proxy or stand-in for other resources in the context of a particular pathway so that relationships between them as components of that pathway can be stated without implying that the those relationships are true of the resources in other contexts.

Add

> **Subject:** ceterms:AssessmentComponent ceterms:BasicComponent ceterms:CocurricularComponent ceterms:CompetencyComponent ceterms:ComponentCondition ceterms:CourseComponent ceterms:CredentialComponent ceterms:ExtracurricularComponent ceterms:JobComponent ceterms:SelectionComponent ceterms:WorkExperienceComponent
> **Predicate:** rdfs:comment 
> **Object:** Instances act as proxies or stand-ins for other resources in the context of a particular pathway so that relationships between them as components of that pathway can be stated without implying that the those relationships are true of the resources in other contexts.

**Comment by stuartasutton** [comment]
I think Phil's revisions make sense. However, I am a bit nervous (and probably over thinking) the use of the word "stand-in" and its potential confusion with the notion of "substitute for"... which pathway components are not.  But I cannot think of any alternatives.

**Comment by siuc-nate** [comment]
>The comment talks about "this proxy resource" which introduces the idea of proxy resources without explaining what it has to do with pathway components, and is about the subject not the use of the property.

Yes, I was deliberately avoiding any mention of pathway-related classes since the scope of this property will no doubt extend beyond them.

--
>Indicates the resource for which a pathway component or similar proxy resource is a stand-in.

Shouldn't we stick to our pattern of "this resource" instead of "a pathway component"? Or at the very least, "**this** pathway component"?

--
>The object for this property will be acting as proxy for the resource that is its subject in a context such as a pathway so that assertions can be made about relationships between things in that context that are not universially true of the resources for which they are proxies.

To me that reads kind of backwards. The "object" (in a subject/predicate/object sense) for the property is the "real" resource, not the proxy.  The subject is the thing (the pathway component in this case) that has the `proxyFor` property.   In other words:
```
[Subject]                                  -[proxyFor]->                [Object]
CredentialComponent                        -[proxyFor]->                Certificate
"...the resource that is its subject..."   -["...this property..."]->   "The object for..."
```

--
>An instance of PathwayComponent may serve as the goal or destination node of a Pathway. A Pathway Component acts as a proxy or stand-in for other resources in the context of a particular pathway so that relationships between them as components of that pathway can be stated without implying that the those relationships are true of the resources in other contexts.

This sounds good, but I would note that it's an update to the `dct:description` ("comment") field rather than the `rdfs:comment` ("definition") field

**Comment by philbarker** [comment]
@siuc-nate 

> To me that reads kind of backwards. The "object" (in a subject/predicate/object sense) for the property is the "real" resource, not the proxy. 

Yes, my mistake. Have fixed the proposal above. Thanks, good catch.

> note that it's an update to the dct:description ("comment") field rather than the rdfs:comment ("definition") field

Again, my mistake. Have fixed the proposal above. Thanks.

On the generalization of the definitions and comment, I think in this case the requirement for a "proxy resource" is quite peculiar, and I don't think we will need them often. I have tried to keep the door open by phrasing as "a pathway component or similar proxy resource". It's already hard enough to explain what's going on without trying to do so without reference to anything specific. In short, I prefer a simpler explanation over something generic.

> Shouldn't we stick to our pattern of "this resource" instead of "a pathway component"? Or at the very least, "this pathway component"?

OK, for the sake of conforming to an established style I have changed to "this pathway component..."

**Comment by siuc-nate** [comment]
> I think in this case the requirement for a "proxy resource" is quite peculiar,

I'm not quite sure what you mean.  The proxy resource _is_ the resource with the `proxyFor` property (i.e. the proxy resource is the pathway component).

**Comment by philbarker** [comment]
> I'm not quite sure what you mean. The proxy resource is the resource with the proxyFor property (i.e. the proxy resource is the pathway component).

Yes. It's not often you need to introduce resources as proxies for something else into the model.

**Comment by siuc-nate** [comment]
We can probably tighten up the comment a little bit:
"Indicates that this resource is a proxy for the referenced resource, so that assertions can be made about that resource that may only be true in this proxy's context."

I'll take a stab at simplifying the proposed addition to the comment for PathwayComponent as well:
"Pathway Component acts as a proxy or stand-in for another resource in the context of a particular pathway so that assertions about that resource can be made that may only be true in the context of that particular pathway."

Likewise for the other components:
"Subclasses of PathwayComoponent component act as a proxy or stand-in for another resource in the context of a particular pathway so that assertions about that resource can be made that may only be true in the context of that particular pathway."

It may be possible to use the same comment for all of the pathway components (not just the subclasses), or it may be preferable to use the name of the particular subclass in place of "Subclasses of PathwayComponent" (e.g. "Credential Component acts as a...", "JobComponent acts as a..." etc.).

**Comment by philbarker** [comment]
> "Indicates that this resource is a proxy for the referenced resource, so that assertions can be made about that resource that may only be true in this proxy's context."

Not bad. But it's the very fact of being a PathwayComponent (or similar) that allows you to make assertions that are only true in context. That would stand whether or not you used `proxyFor`.

I'm having second thoughts about talking about "**this** resource" when defining a property. Which resource: the object? the subject? the property? 

I prefer
> **Definition:** Indicates the resource for which a pathway component or similar proxy resource is a stand-in.

> "Pathway Component acts as a proxy or stand-in for another resource in the context of a particular pathway so that assertions about that resource can be made that may only be true in the context of that particular pathway."

PathwayComponent is a class, and as such not a stand-in for another resource. It's instances of the class that are the stand-ins. Same for subclasses. Better to say "Pathway components act as proxies or ..." but that doesn't allow for the same description text on all the subclasses. Hence my use of "instances".

I'm not sure if you are suggesting dropping the "An instance of PathwayComponent may serve as the goal or destination node of a Pathway." part of the description, which is what currently exists.

I think we are describing the class, and what it does, rather than telling people how to use it.

**Comment by siuc-nate** [comment]
>Not bad. But it's the very fact of being a PathwayComponent (or similar) that allows you to make assertions that are only true in context. That would stand whether or not you used `proxyFor`.

If that's true, then do we actually need the proposed change to the comment for Pathway Component at all?  It seems like we would be better off not saying it if we don't need to.  I think having the proposed comment on `proxyFor` is probably enough.

--

> I'm having second thoughts about talking about "**this** resource" when defining a property. Which resource: the object? the subject? the property?

"this resource" is a pattern we use throughout CTDL to mean "the thing that is doing the pointing/has the property", and we use "referenced resource" at least twice in CTDL-ASN to indicate "the thing this property is pointing to".  However, our use/phrasing of "instances"/"an instance" is much less consistent.
![image](https://user-images.githubusercontent.com/21346519/118997622-6a7b8680-b94e-11eb-88d1-a43867ec7073.png)

--

> Indicates the resource for which a pathway component or similar proxy resource is a stand-in.

I could similarly ask, "**A** pathway component? Which one?"  And again, I don't want to restrict this property to just pathway components. 

--

> PathwayComponent is a class, and as such not a stand-in for another resource. It's instances of the class that are the stand-ins.

Fair point. 

> I'm not sure if you are suggesting dropping the "An instance of PathwayComponent may serve as the goal or destination node of a Pathway." part of the description, which is what currently exists.

Not intentionally, though I suppose that is a comment whose information could be inferred from looking at the domains and ranges of `ceterms:hasDestinationComponent` and `ceterms:isDestinationComponentOf`.  But now that you mention it (existing comments, that is), some of the Pathway Component subclasses already have comments, which we would also need to write around.  

Perhaps the proposed comment (assuming we still need it at all, that is) really only needs to exist on the Pathway Component itself ?  Since the others are subclasses of it, it's reasonable to assume (until someone complains) they would share its nature I think.  Then we could be a bit more specific:
"An instance of a Pathway Component subclass may serve as the goal or destination node of a Pathway and/or may act as a proxy or stand-in for another resource so that assertions can be made about that resource that may only be true in the context of a Pathway. "

We do use "and/or" a number of places as well:
![image](https://user-images.githubusercontent.com/21346519/119002905-e4ae0a00-b952-11eb-9e56-1badf1efe501.png)

**Comment by philbarker** [comment]
> do we actually need the proposed change to the comment for Pathway Component at all? 

I think it would be helpful to explain what a class is for in the definition of the class rather than in the definition of one of the properties (or of its superclass).

> I could similarly ask, "A pathway component? Which one?"

The difference is that `proxyFor` will be used in triples with only one pathway component (or similar proxy) whereas the subject, object and property are all RDF Resources. It's kind of odd to define an RDF resource with phrasing of "this resource" and not mean the resource you're defining.  I understand that the "this resource" phrasing is a pattern, but it seems to me to come from thinking in terms of OO classes and instances encapsulating properties rather than from RDF. Sometimes breaking from a pattern is an improvement.

**Comment by siuc-nate** [comment]
>I think it would be helpful to explain what a class is for in the definition of the class rather than in the definition of one of the properties (or of its superclass).

But you indicated that (emphasis mine):
>**the very fact of being a PathwayComponent** (or similar) that allows you to make assertions that are only true in context. That would stand whether or not you used `proxyFor`.

Which would seem to imply that, simply _because_ it is a Pathway Component, you already know that it is a proxy for something else, regardless of `proxyFor` (and thus don't need that information laid out in a comment).  Or am I misinterpreting your earlier remark?  I guess it's not a big deal either way, but you had previously mentioned RDF erring on the side of not saying something, so it seemed to me that if we didn't need to add a comment to Pathway Component, we may as well not do so.  If that's not the case, then let's proceed.

--

>It's kind of odd to define an RDF resource with phrasing of "this resource" and not mean the resource you're defining.

But I do mean the resource I'm defining:
```
("this resource") -[proxyFor]-> ("referenced resource")
```
Or, more grammatically:
```
("this resource") -[is a proxyFor]-> ("that resource")
```
The starting point (subject) being local ("this") and the destination (object) being somewhere else ("that") - it seems to make perfect sense to me.  It's based in english sentence structure, not (just) in OO thinking, and follows how the rest of CTDL works.  In my opinion, it is more intuitive than:
```
("a pathway component") -[proxyFor]-> ("the resource")
```
Maybe someone else can chime in to give their thoughts? I feel like we're spinning our wheels a bit.

--

>but it seems to me to come from thinking in terms of OO classes and instances encapsulating properties rather than from RDF

There is a bit of that, yes, but again, that's how the rest of CTDL works, and those kinds of patterns help make the schema predictable, easier to understand, and more intuitive.  It's especially important when working with so many developers (most if not all of whom are OO thinkers) that we have a way to bridge the gap between an RDF mindset and an OO mindset.  I think treating the subject as the "thing" that has the properties is perfectly acceptable (and it is how JSON-LD works after all), especially given the lack of a good alternative in terms of programming languages (at least that I'm aware of).

I do agree with you that breaking patterns sometimes is how you evolve, but you have to weigh that against the overhead incurred in being able to explain why something is "this way here and that way there" and the various conditional "if/else/and/or/but/not"s it introduces in the documentation, in code workarounds, in validation/debugging, etc. Sometimes it's worth it.  In this case, I don't think we gain anything by replacing "this" with "a" and "referenced" with "the" (even if we made that change across the entire schema for the sake of pattern consistency).  I think we would lose clarity if we did so.

But again, if someone else wants to chime in, maybe a third perspective will be useful to us both.

**Comment by philbarker** [comment]
> But you indicated that (emphasis mine):
> 
> > **the very fact of being a PathwayComponent** (or similar) that allows you to make assertions that are only true in context. That would stand whether or not you used proxyFor.
> 
> Which would seem to imply that, simply because it is a Pathway Component, you already know that it is a proxy for something else, regardless of proxyFor (and thus don't need that information laid out in a comment).

My thinking is that other people won't know that unless we tell them. 

---

> >It's kind of odd to define an RDF resource with phrasing of "this resource" and not mean the resource you're defining.
> 
> But I do mean the resource I'm defining:
> ```
> ("this resource") -[proxyFor]-> ("referenced resource")
> ```

But here we are discussing how to define the property proxyFor, which, being an RDF property, is a resource, so we are saying
```
proxyFor a rdf:Property .   # which entails <proxyFor> a rdfs:Resource
proxyFor dct:description "this resource...." .
```

And yes, everyone, other perspectives would help.

**Comment by stuartasutton** [comment]
Guys, it would help me (and might clarify for others), if **one of you** did a concise statement (without lots of narrative or justification) of the following: 

- This is what I am hearing you say it should be; and
- This is what I am saying it should be

**Comment by siuc-nate** [comment]
@philbarker 
> But here we are discussing how to define the property proxyFor, which, being an RDF property, is a resource, so we are saying

That's true in a pure RDF sense but it breaks the pattern we've established in CTDL. My stance here is that it is also far more intuitive to define the property in terms of the objects it touches where the subject is the "local" host ("this resource") and the object is the "over there" thing ("referenced resource"). 

Look at almost any property in CTDL that connects two primary objects together and you will find a "this resource" or "this credential" or "this [whatever]", with a few exceptions that might use "the" instead of "this" to refer to the subject of the triple.  If we're going to go the pure RDF route then we need to update an awful lot of definitions across the schema.

We will confuse object-oriented developers if we suddenly start referring to properties themselves as resources even if it may be technically true in a pure RDF sense, especially if it makes it harder to understand what the function of the property actually is. Properties are predicates in triples, triples have a subject and an object, the subject hosts the property to point to the object. It's simple, it's intuitive, it matches what the JSON-LD looks like, it matches what the Turtle looks like - I don't see (and this is maybe where additional perspectives can help) how all of that is outweighed by being able to use a pure RDF-oriented definition for a property that just simply needs to say "this stands in for that" instead of the more ambiguous "something stands in for something else". Where's the benefit?

@stuartasutton 
I think part of the confusion also has to do with this issue sort of conflating the nature of the proposed proxyFor property with the idea of adding similarly-themed comment(s) to PathwayComponent and/or its subclasses.

Phil, would you like to take a stab at Stuart's request?

**Comment by philbarker** [comment]
Summary: my proposal adapted from [comment](https://github.com/CredentialEngine/Schema-Development/issues/769#issuecomment-844000431) above with what I am hearing @siuc-nate say inline:

> **URI:** ceterms:proxyFor
> **Label:** Proxy For
> **Definition:** Indicates the resource for which a pathway component or similar proxy resource is a stand-in.
> **Comment:** The subject for this property will be acting as proxy in a context such as a pathway for the resource that is its object, so that assertions can be made about relationships between things in that context that are not universially true of the resources for which they are proxies.

Nate would prefer the definition to be "Indicates this resource is a stand-in for the referenced resource"

> **Remove**
> 
> > **Subject:** ceterms:PathwayComponent
> > **Predicate:** dct:description
> > **Object:** An instance of PathwayComponent may serve as the goal or destination node of a Pathway.
> 



> **Add**
> 
> > **Subject:** ceterms:PathwayComponent
> > **Predicate:** dct:description 
> > **Object:** An instance of PathwayComponent may serve as the goal or destination node of a Pathway. A Pathway Component acts as a proxy or stand-in for other resources in the context of a particular pathway so that relationships between them as components of that pathway can be stated without implying that the those relationships are true of the resources in other contexts.
> 
> Add
> 
> > **Subject:** ceterms:AssessmentComponent ceterms:BasicComponent ceterms:CocurricularComponent ceterms:CompetencyComponent ceterms:ComponentCondition ceterms:CourseComponent ceterms:CredentialComponent ceterms:ExtracurricularComponent ceterms:JobComponent ceterms:SelectionComponent ceterms:WorkExperienceComponent
> > **Predicate:** dct:description 
> > **Object:** Instances act as proxies or stand-ins for other resources in the context of a particular pathway so that relationships between them as components of that pathway can be stated without implying that the those relationships are true of the resources in other contexts.
> 

I think we are both equivocal about whether to keep the "An instance of PathwayComponent may serve as the goal or destination node of a Pathway"

I think Nate questions whether the additonal description is needed at all.

**Comment by stuartasutton** [comment]
I would accept these four changes as-is.  I think the `proxyFor` definition is very precise; and, I would not be concerned with deviations from the "this resource"/"referenced resource" pattern. 

I find the "truth" statement useful. I do not believe that we can assume that most developers would otherwise gather that the assertion could be assumed without the explicit "truth" statement.  Better to state than rely on an assumption.

I think this proposal is ready to go.

**Comment by siuc-nate** [comment]
>I would not be concerned with deviations from the "this resource"/"referenced resource" pattern.

Can you elaborate on this? I thought sticking to patterns was very important.  It certainly makes it easier to predict or intuit how the schema works.

**Comment by stuartasutton** [comment]
Yes, a total lack of patterns is not a good idea. But slavish adherence to
patterns when the occasional alternative presents a more meaningful or
useful result is, as Ralph Waldo Emerson once said, the "hobgoblin of
little minds".

**Comment by siuc-nate** [comment]
>when the occasional alternative presents a more meaningful or useful result

I agree, when the alternative is more meaningful or useful - but my position is that:
- "a" (as in "a pathway component...") is less specific (less meaningful, less useful) than "this" (as in "this pathway component...")
- "the" as in ("the resource for which") is equally specific as "referenced" (as in "the referenced resource") except that everywhere else I could find in CTDL where we say "the resource", we mean the subject, not the object, of the triple where the property is the predicate.  That would make this definition both less specific than other definitions and make it read backwards compared to them.

If those two things presented a result that was more meaningful or useful, I would be on board with it and probably recommend we change the rest of CTDL to match - but I don't see how we benefit from doing things that way here.  That is what I was hoping you would elaborate on: what is the benefit of breaking the pattern here?

**Comment by philbarker** [comment]
@siuc-nate we seem to have neglected to state the obvious: that the domain and range of `ceterms:proxyFor` should be the same as `ceterms:sourceData`

**Comment by stuartasutton** [comment]
Nothing precludes there being more than one comment, is there?
